diff --git a/examples/step-93/step-93.cc b/examples/step-93/step-93.cc
index f7b73cfd24..cd3392a9a5 100644
--- a/examples/step-93/step-93.cc
+++ b/examples/step-93/step-93.cc
@@ -312,7 +312,7 @@ namespace Step93
   void Step93<dim>::make_grid()
   {
     GridGenerator::hyper_cube(triangulation, -1, 1);
-    triangulation.refine_global(7);
+    triangulation.refine_global(3);
 
     std::cout << "Number of active cells: " << triangulation.n_active_cells()
               << std::endl;
@@ -642,8 +642,8 @@ namespace Step93
   template <int dim>
   void Step93<dim>::solve()
   {
-    Timer timer;
-    timer.start();
+    //Timer timer;
+    //timer.start();
 
     std::cout << "Beginning solve..." << std::endl;
 
@@ -652,9 +652,9 @@ namespace Step93
 
     solver.solve(system_matrix, solution, system_rhs, PreconditionIdentity());
 
-    timer.stop();
+    //timer.stop();
 
-    std::cout << "Wall time: " << timer.wall_time() << "s" << std::endl;
+    //std::cout << "Wall time: " << timer.wall_time() << "s" << std::endl;
     std::cout << "Solved in " << solver_control.last_step()
               << " MINRES iterations." << std::endl;
   }
